### PR TITLE
S3 Sharing: Modifications to LaraProvider, Sharing Dialog, AutoLaunch page

### PR DESCRIPTION
**Pull request by knowuh**
AutoLaunching of shared DocumentStore CFM documents were handled by a special endpoint in the Document server.

This PR sets up a new autolaunching URL hosted at `https://cloud-file-manger.concord.org/<branch-deploy-dir>//autolaunch/autolaunch.html`

* The FileSharing dialog had to be modified to accommodate the variable autolaunch page.
* The Lara Provider had to be configured to look for S3 public document URLS, and copy those documents into student work (using same old docstore + LARA interactive runtime state scheme).

#173017375

### CODE REVIEW

File: src/code/views/share-dialog-view.js [hunk 1]
```diff
@@ -63,23 +64,34 @@ export default createReactClass({
       pageType: "autolaunch",
       serverUrl: this.props.settings.serverUrl || "https://codap.concord.org/releases/latest/",
       serverUrlLabel: this.props.settings.serverUrlLabel || translate("~SHARE_DIALOG.LARA_CODAP_URL"),
+
+      // TODO: Maybe change the defaults to point at http://cloud-file-manager.concord.org/?
```
Comment (dougmartin) [inline src/code/views/share-dialog-view.js:68]: TODONE

File: src/code/views/share-dialog-view.js [hunk 2]
```diff
@@ -117,6 +129,38 @@ export default createReactClass({
...
+    // <AutoLaunchPageUrl> expects a `documentId` param (was a path in DocStore)
+    //    and a `server` param, that points usually to some CODAP release.
+    // <CODAP SERVER> can have url encoded params too such as `FcfmBaseUrl=`
+    // where URL is the url to the /js folder for CFM
+    //
+    // It will point to SPA hosted in this repo -- NP 2020-06-29
+    //  1: Get the resource URL (S3 shared document public URL)
+    //  2: Get the URL for the autoLauch page (hosted here ...)
+    //  3: Construct a URL to <AutlaunchPage
+    // TODO: What happens if we are launched from CFM build inside CODAP?
```
Comment (dougmartin) [inline src/code/views/share-dialog-view.js:147]: Get rid of this?

File: src/code/views/share-dialog-view.js [hunk 3]
```diff
@@ -117,6 +129,38 @@ export default createReactClass({
...
+    const encodedCFMBase = encodeURIComponent(localCFMBaseUrl)
+
+    const documentServer = encodeURIComponent(this.getShareLink())
+    const graphVisToggles = this.state.graphVisToggles ? '&app=is' : ''
+    // graphVisToggles is a parameter handled by CODAP, so it needs to be added to its URL.
+    const server = encodeURIComponent(`${this.state.serverUrl}?cfmBaseUrl=${encodedCFMBase}${graphVisToggles}`)
+    // Other params are handled by document server itself:
+    const buttonText = this.state.pageType === 'launch' ? `&buttonText=${encodeURIComponent(this.state.launchButtonText)}` : ''
+    const fullscreenScaling = (this.state.pageType === 'autolaunch') && this.state.fullscreenScaling ? '&scaling' : ''
+
```
Comment (dougmartin) [inline src/code/views/share-dialog-view.js:160]: Maybe add TODO to use queryString if this code is touched again, maybe...????

**Comment by knowuh** [comment]
Verified that the Select linked state dialog box was working:

Screen Shot 2020-07-01 at 3 46 56 PM

**Comment by knowuh** [comment]
@scytacki  @dougmartin Thanks for the review.

I found a few other issues when addressing them.  I also verified that choosing linked run states worked.  The HTML of autolaunch.html was complete garbage, and included `<image />` tags instead of `` tags and other horrors. That was fun. :)

All the changes  in ha75f341  :

Fixed:
* HTML formatting.  In addtion to the PR requests, HTML vlidator found <image> tags (not ) plus others.
* Wait until document loaded before calling autoLaunchInteractive()
* Handle (+ verified) cases where multiple linked run states exist and must be selected.
* Removed 2 TODONES and added one TODO
* Moved type declarations from dependencies to devDependencies in package.json

**Comment by dougmartin** [comment]
Looks good :+1:

**Comment by scytacki** [comment]
One thing I just remembered is that this shouldn't make a new document for the student unless they actually change something in CODAP. At least that is how I think it used to work. I'm going review the code again to check for that case.

**Comment by scytacki** [comment]
As far as I can tell the LARA provider and the autolaunch has always made new documents as soon as the student loads them in LARA. So this is doing the same as before. I believe authors are using the LARA click to play feature to prevent the document from loading until the student actually clicks on the interactive.

**Pull request merged**